openhcl/page_pool_alloc/src/lib.rs

```diff
@@ -17,12 +17,161 @@ use anyhow::Context;
 use hcl::ioctl::MshvVtlLow;
 use hvdef::HV_PAGE_SIZE;
 use inspect::Inspect;
+use memory_range::MemoryRange;
 use parking_lot::Mutex;
 use std::num::NonZeroU64;
 use std::sync::Arc;
 use thiserror::Error;
 use vm_topology::memory::MemoryRangeWithNode;
 
+/// Save restore suport for [`PagePool`].
+pub mod save_restore {
+    use super::PagePool;
+    use super::Slot;
+    use super::SlotState;
+    use memory_range::MemoryRange;
+    use mesh::payload::Protobuf;
+    use vmcore::save_restore::SaveRestore;
+    use vmcore::save_restore::SavedStateRoot;
+
+    #[derive(Protobuf)]
+    #[mesh(package = "openhcl.pagepool")]
+    enum InnerSlotState {
+        #[mesh(1)]
+        Free,
+        #[mesh(2)]
+        Allocated {
+            #[mesh(1)]
+            device_id: String,
+            #[mesh(2)]
+            tag: String,
+        },
+        #[mesh(3)]
+        Leaked {
+            #[mesh(1)]
+            device_id: String,
+            #[mesh(2)]
+            tag: String,
+        },
+    }
+
+    #[derive(Protobuf)]
+    #[mesh(package = "openhcl.pagepool")]
+    struct SlotSavedState {
+        #[mesh(1)]
+        base_pfn: u64,
+        #[mesh(2)]
+        size_pages: u64,
+        #[mesh(3)]
+        state: InnerSlotState,
+    }
+
+    /// The saved state for [`PagePool`].
+    #[derive(Protobuf, SavedStateRoot)]
+    #[mesh(package = "openhcl.pagepool")]
+    pub struct PagePoolState {
+        #[mesh(1)]
+        state: Vec<SlotSavedState>,
+        #[mesh(2)]
+        ranges: Vec<MemoryRange>,
+        #[mesh(3)]
+        pfn_bias: u64,
```

Comment:
Do we want/need to save this?

Reply:
hmm... I think it's useful for sanity checking you didn't restore say the shared pages into the non-shared but perhaps not if we saved the pool typ, which we don't either. Given we already save the ranges, I 

TBH not sure if it really matters since we'll save/restore the whole partition's address space map, so I think i'll just remove it.